File: TTR/trunk/src/runFun.f
c-----------------------------------------------------------------------c
c TTR, copyright (C) Joshua M. Ulrich, 2007                             c
c Distributed under GNU GPL version 3                                   c
c-----------------------------------------------------------------------c

c
c     Calculate a running/rolling SUM
c
c     ia  : input array
c     lia : length of input array
c     n   : size of window
c     oa  : output array
c     loa : length of output array
c
      subroutine runsum(ia, lia, n, oa, loa)

      integer lia, n, loa, i
      double precision ia(lia), oa(loa)

      do 10 i=n+1,lia

        oa(i) = oa(i-1) + ia(i) - ia(i-n)

   10 continue
      end
c
c     Calculate a Welles Wilder SUM
c
c     ia  : input array
c     lia : length of input array
c     n   : size of window
c     oa  : output array
c     loa : length of output array
c
      subroutine wilder(ia, lia, n, oa, loa)

      integer lia, n, loa, i
      double precision ia(lia), oa(loa)

      do 10 i=2,lia

        oa(i) = ia(i) + oa(i-1) * (n-1)/n

   10 continue
      end
c
c     Calculate a running/rolling MIN
c     (Not very elegant, but _much_ faster
c     than a pure R implementation)
c
c     ia   : input array
c     lia  : length of input array
c     n    : size of window
c     oa   : output array
c     loa  : length of output array
c     lmin : window minimum
c
      subroutine runmin(ia, lia, n, oa, loa)

      integer lia, n, loa, i, j
      double precision ia(lia), oa(loa), lmin

      do 10 i=n,lia
        
        lmin = ia(i)

        do 20 j=i-n+1,i-1
          
          lmin = min( ia(j), lmin )

   20 continue

        oa(i) = lmin

   10 continue
      end
c
c     Calculate a running/rolling MAX
c     (Not very elegant, but _much_ faster
c     than a pure R implementation)
c
c     ia   : input array
c     lia  : length of input array
c     n    : size of window
c     oa   : output array
c     loa  : length of output array
c     lmax : window maximum
c
      subroutine runmax(ia, lia, n, oa, loa)

      integer lia, n, loa, i, j
      double precision ia(lia), oa(loa), lmax

      do 10 i=n,lia
        
        lmax = ia(i)

        do 20 j=i-n+1,i-1
          
          lmax = max( ia(j), lmax )

   20 continue

        oa(i) = lmax

   10 continue
      end
c
c     Calculate a running/rolling SUM of PRODUCTS
c     (Not very elegant, but _much_ faster
c     than a pure R implementation)
c
c     ia1  : input array #1
c     ia2  : input array #2
c     lia  : length of input arrays
c     n    : size of window
c     oa   : output array
c     loa  : length of output array
c     prod : window product
c
      subroutine runSP(ia1, ia2, lia, n, oa, loa)

      integer lia, n, loa, i, j
      double precision ia1(lia), ia2(lia), oa(loa)
      
      do 10 i=n,lia

        prod = ia1(i) * ia2(i)

        do 20 j=i-n+1,i-1
          
          prod = prod + ( ia1(j) * ia2(j) )

   20 continue

        oa(i) = prod

   10 continue
      end

File: TTR/trunk/R/runFun.R
#-------------------------------------------------------------------------#
# TTR, copyright (C) Joshua M. Ulrich, 2007                               #
# Distributed under GNU GPL version 3                                     #
#-------------------------------------------------------------------------#

"runSum" <-
function(x, n=10) {

  x   <- as.vector(x)

  if( n < 1 | n > NROW(x) ) stop("Invalid 'n'")

  # Count NAs, ensure they're only at beginning of data, then remove.
  NAs <- sum( is.na(x) )
  if( NAs > 0 ) {
    if( any( is.na(x[-(1:NAs)]) ) ) stop("Series contains non-leading NAs")
  }
  x   <- na.omit(x)

  # Initialize result vector 
  result <- rep(0,NROW(x))
  result[n] <- sum(x[1:n])

  # Call Fortran routine
  result <- .Fortran( "runsum",
                   ia = as.double(x),
                   lia = as.integer(NROW(x)),
                   n = as.integer(n),
                   oa = as.double(result),
                   loa = as.integer(NROW(result)),
                   PACKAGE = "TTR" )$oa

  # Replace 1:(n-1) with NAs and prepend NAs from original data
  result[1:(n-1)] <- NA
  result <- c( rep( NA, NAs ), result )

  return( result )
}

#-------------------------------------------------------------------------#

"wilderSum" <-
function(x, n=10) {

  x   <- as.vector(x)

  if( n < 1 | n > NROW(x) ) stop("Invalid 'n'")

  # Count NAs, ensure they're only at beginning of data, then remove.
  NAs <- sum( is.na(x) )
  if( NAs > 0 ) {
    if( any( is.na(x[-(1:NAs)]) ) ) stop("Series contains non-leading NAs")
  }
  x   <- na.omit(x)

  # Initialize result vector 
  result <- x

  result <- .Fortran( "wilder",
                   ia  = as.double(x),
                   lia = as.integer(NROW(x)),
                   n   = as.integer(n),
                   oa  = as.double(result),
                   loa = as.integer(NROW(result)),
                   PACKAGE = "TTR" )$oa

  # Replace 1:(n-1) with NAs and prepend NAs from original data
  result[1:(n-1)] <- NA
  result <- c( rep( NA, NAs ), result )

  return( result )
}

#-------------------------------------------------------------------------#

"runMin" <-
function(x, n=10) {

  x   <- as.vector(x)

  if( n < 1 | n > NROW(x) ) stop("Invalid 'n'")

  # Count NAs, ensure they're only at beginning of data, then remove.
  NAs <- sum( is.na(x) )
  if( NAs > 0 ) {
    if( any( is.na(x[-(1:NAs)]) ) ) stop("Series contains non-leading NAs")
  }
  x   <- na.omit(x)

  # Initialize result vector 
  result <- rep(0,NROW(x))
  result[n] <- min(x[1:n])

  result <- .Fortran( "runmin",
                   ia = as.double(x),
                   lia = as.integer(NROW(x)),
                   n = as.integer(n),
                   oa = as.double(result),
                   loa = as.integer(NROW(result)),
                   PACKAGE = "TTR" )$oa

  # Replace 1:(n-1) with NAs and prepend NAs from original data
  result[1:(n-1)] <- NA
  result <- c( rep( NA, NAs ), result )

  return( result )
}

#-------------------------------------------------------------------------#

"runMax" <-
function(x, n=10) {

  x   <- as.vector(x)
  
  if( n < 1 | n > NROW(x) ) stop("Invalid 'n'")

  # Count NAs, ensure they're only at beginning of data, then remove.
  NAs <- sum( is.na(x) )
  if( NAs > 0 ) {
    if( any( is.na(x[-(1:NAs)]) ) ) stop("Series contains non-leading NAs")
  }
  x   <- na.omit(x)

  # Initialize result vector 
  result <- rep(0,NROW(x))
  result[n] <- max(x[1:n])

  result <- .Fortran( "runmax",
                   ia = as.double(x),
                   lia = as.integer(NROW(x)),
                   n = as.integer(n),
                   oa = as.double(result),
                   loa = as.integer(NROW(result)),
                   PACKAGE = "TTR" )$oa

  # Replace 1:(n-1) with NAs and prepend NAs from original data
  result[1:(n-1)] <- NA
  result <- c( rep( NA, NAs ), result )

  return( result )
}

#-------------------------------------------------------------------------#

"runVar" <-
function(x, n=10, sample=TRUE) {

  x <- as.double( as.vector(x) )

  result <- runSum(x*x, n)/n - (runSum(x, n)/n)^2

  if(sample) {
    result <- result * n/(n+1)
  }

  return(result)
}

#-------------------------------------------------------------------------#

"runMedian" <-
function(x, n=10, low=FALSE, high=FALSE) {

  x   <- as.vector(x)

  if( n < 1 | n > NROW(x) ) stop("Invalid 'n'")

  # Count NAs, ensure they're only at beginning of data, then remove.
  NAs <- sum( is.na(x) )
  if( NAs > 0 ) {
    if( any( is.na(x[-(1:NAs)]) ) ) stop("Series contains non-leading NAs")
  }
  x   <- na.omit(x)

  # Initialize result vector 
  result <- rep(0,NROW(x))

  # Call Fortran routine
  result <- .Fortran( "runmedian",
                   ia = as.double(x),
                   n = as.integer(n),
                   oa = double(NROW(x)),
                   la = as.integer(NROW(x)),
                   ver = as.integer(version),
                   PACKAGE = "TTR" )$oa

  # Replace 1:(n-1) with NAs and prepend NAs from original data
  result[1:(n-1)] <- NA
  result <- c( rep( NA, NAs ), result )

  return( result )
}
#-------------------------------------------------------------------------#

"runMAD" <-
function(x, n=10, center=runMedian(x, n), stat="median",
         constant=1.4826, low=FALSE, high=FALSE) {

  x   <- as.vector(x)

  if( n < 1 | n > NROW(x) ) stop("Invalid 'n'")

  # Count NAs, ensure they're only at beginning of data, then remove.
  NAs <- sum( is.na(x) )
  if( NAs > 0 ) {
    if( any( is.na(x[-(1:NAs)]) ) ) stop("Series contains non-leading NAs")
  }
  x   <- na.omit(x)
  center[1:(n-1)] <- 0

  # Initialize result vector 
  result <- rep(0,NROW(x))

  # Mean or Median absolute deviation?
  median <- match.arg(stat, c("mean","median"))
  median <- switch( stat, median=TRUE, mean=FALSE )

  # 'version' = f( low, high )
  if( low && high )
    stop("'low' and 'high' cannot be both TRUE")
  version <- 0
  if( low )  version <- -1
  if( high )  version <- 1


  # Call Fortran routine
  result <- .Fortran( "runMAD",
                   rs = as.double(x),
                   cs = as.double(center),
                   la = as.integer(NROW(x)),
                   n = as.integer(n),
                   oa = as.double(result),
                   stat = as.integer(median),
                   ver = as.integer(version),
                   PACKAGE = "TTR" )$oa

  if( median ) result <- result * constant

  # Replace 1:(n-1) with NAs and prepend NAs from original data
  result[1:(n-1)] <- NA
  result <- c( rep( NA, NAs ), result )

  return( result )
}
